REGRESSION (271515@main): Bold and Italic keyboard bar button items in the UCB never update

whsieh opened the pull request:
#### 9e343906838f775714988ac2c3d6be7d7ef157a3
<pre>
REGRESSION (271515@main): Bold and Italic keyboard bar button items in the UCB never update
<a href="https://bugs.webkit.org/show_bug.cgi?id=270169">https://bugs.webkit.org/show_bug.cgi?id=270169</a>
<a href="https://rdar.apple.com/123328041">rdar://123328041</a>

Reviewed by Tim Horton.

The changes in 271515@main were meant to merge functionality in both the public API delegate method
`-textStylingAtPosition:inDirection:` and the UIKit IPI `-fontForCaretSelection` into just the
public API. However, this removed logic for setting symbolic traits on the font descriptor (bold,
italic) with no replacement.

Fix this by reinstating this code, and augmenting an existing API test to exercise this logic.

Test: AutocorrectionTests.FontAtCaretWhenUsingUICTFontTextStyle

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textStylingAtPosition:inDirection:]):
* Tools/TestWebKitAPI/Tests/ios/AutocorrectionTestsIOS.mm:
(-[UIFont _test_systemFontWithSize:traits:]):

Canonical link: <a href="https://commits.webkit.org/275396@main">https://commits.webkit.org/275396@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f6aa734d537bd29d079f342d3b9274151aa8ca7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41735 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20749 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44117 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44311 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37828 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44042 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23886 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18080 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34495 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17681 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35945 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15165 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15373 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36969 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45700 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37925 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37289 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41035 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16550 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13587 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39470 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18169 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9354 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18226 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17813 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -4263,15 +4263,28 @@ - (NSDictionary *)textStylingAtPosition:(UITextPosition *)position inDirection:(
+            newTraits |= UIFontDescriptorTraitItalic;
+
+        if (originalTraits != newTraits) {
+            RetainPtr descriptor = [[font fontDescriptor] ?: adoptNS([UIFontDescriptor new]) fontDescriptorWithSymbolicTraits:newTraits];
```
What's happening in this line 😅

whsieh replied:
We're using the `font`'s `fontDescriptor` (if it exists), and falling back to a new `UIFontDescriptor` instance if it doesn't exist.

aprotyas replied:
Makes sense, I totally didn't realize you could use the ternary operator like that.

whsieh committed:
REGRESSION (271515@main): Bold and Italic keyboard bar button items in the UCB never update
https://bugs.webkit.org/show_bug.cgi?id=270169
rdar://123328041

Reviewed by Tim Horton.

The changes in 271515@main were meant to merge functionality in both the public API delegate method
`-textStylingAtPosition:inDirection:` and the UIKit IPI `-fontForCaretSelection` into just the
public API. However, this removed logic for setting symbolic traits on the font descriptor (bold,
italic) with no replacement.

Fix this by reinstating this code, and augmenting an existing API test to exercise this logic.

Test: AutocorrectionTests.FontAtCaretWhenUsingUICTFontTextStyle

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView textStylingAtPosition:inDirection:]):
* Tools/TestWebKitAPI/Tests/ios/AutocorrectionTestsIOS.mm:
(-[UIFont _test_systemFontWithSize:traits:]):

Canonical link: https://commits.webkit.org/275396@main
